### DISCUSSION TITLE
[Deployment] zkSync era testnet randomly gives `not enough gas to publish compressed bytecodes`

### DISCUSSION THREAD

**Discussion by PatrickAlphaC** [zkSync-Community-Hub/zksync-developers#550] [Deployments] [answered]
### Environment

Testnet

### zkSolc Version

1.4.1

### zksync-ethers Version

6.7.0

### Hardhat.config.ts

```javascript
import { HardhatUserConfig } from "hardhat/config"

import "@matterlabs/hardhat-zksync-node"
import "@matterlabs/hardhat-zksync-deploy"
import "@matterlabs/hardhat-zksync-solc"
import "@matterlabs/hardhat-zksync-verify"
import "@nomicfoundation/hardhat-toolbox"

const config: HardhatUserConfig = {
  networks: {
    zkSyncSepoliaTestnet: {
      url: process.env.ZKSYNC_SEPOLIA_RPC_URL!,
      ethNetwork: "sepolia",
      zksync: true,
      verifyURL: "https://explorer.sepolia.era.zksync.dev/contract_verification",
      accounts: [process.env.ZKSYNC_SEPOLIA_KEY!],
    },
    zkSyncMainnet: {
      url: process.env.ZKSYNC_RPC_URL!,
      ethNetwork: "mainnet",
      zksync: true,
      verifyURL: "https://zksync2-mainnet-explorer.zksync.io/contract_verification",
      accounts: [process.env.ZKSYNC_KEY!],
    },
    zkSyncGoerliTestnet: { // deprecated network
      url: "https://testnet.era.zksync.dev",
      ethNetwork: "goerli",
      zksync: true,
      verifyURL: "https://zksync2-testnet-explorer.zksync.dev/contract_verification",
    },
    dockerizedNode: {
      url: "http://localhost:3050",
      ethNetwork: "http://localhost:8545",
      zksync: true,
    },
    inMemoryNode: {
      url: "http://127.0.0.1:8011",
      ethNetwork: "localhost", // in-memory node doesn't support eth node; removing this line will cause an error
      zksync: true,
    },
    hardhat: {
      zksync: true,
    },
  },
  zksolc: {
    version: "latest",
    settings: {
      // find all available options in the official documentation
      // https://era.zksync.io/docs/tools/hardhat/hardhat-zksync-solc.html#configuration
    },
  },
  solidity: {
    version: "0.8.24",
  },
}

export default config
```

### Deployment Script (WITHOUT PRIVATE KEY)

```javascript
import { deployContract, DeployContractOptions, getWallet } from "./utils"
import * as zk from 'zksync-ethers'

export default async function deploy(): Promise<zk.Contract> {
    const result = await deploySecurityRegistry()
    return result
}

export async function deploySecurityRegistry(): Promise<zk.Contract> {
    let wallet = getWallet()
    let deployOptions: DeployContractOptions | undefined = { wallet: wallet }
    /*//////////////////////////////////////////////////////////////
                           Registry
    //////////////////////////////////////////////////////////////*/
    const securityCourseRegistry: string = "CyfrinSecurityCourseNft"
    const securityCourseArgs: string[] = []
    // This will also verify
    const securityCourseContract = await deployContract(securityCourseRegistry, securityCourseArgs, deployOptions)
    const securityCourseAddress = await securityCourseContract.getAddress()
    console.log(`Deployed registry contract to: ${securityCourseAddress}`)
    return securityCourseContract
}
```
```


### Package.json

```json
{
  "name": "nfts-zksync-hardhat",
  "description": "A template for zkSync smart contracts development with Hardhat",
  "private": true,
  "author": "Cyfrin",
  "license": "MIT",
  "scripts": {
    "deploy": "hardhat deploy-zksync --script deploy.ts",
    "interact": "hardhat deploy-zksync --script interact.ts",
    "compile": "hardhat compile",
    "clean": "hardhat clean",
    "test": "hardhat test --network hardhat"
  },
  "devDependencies": {
    "@types/chai": "^4.3.4",
    "@types/mocha": "^10.0.1",
    "chai": "^4.3.7",
    "dotenv": "^16.0.3",
    "ethers": "^6.9.2",
    "hardhat": "^2.12.4",
    "mocha": "^10.2.0",
    "ts-node": "^10.9.1",
    "typescript": "^4.9.5",
    "zksync-ethers": "^6.7.0"
  },
  "dependencies": {
    "@chainlink/contracts": "^1.1.1",
    "@matterlabs/hardhat-zksync": "^1.0.0",
    "@matterlabs/hardhat-zksync-ethers": "^1.0.0",
    "@matterlabs/hardhat-zksync-verify": "^1.4.3",
    "@nomicfoundation/hardhat-ethers": "^3.0.6",
    "@nomicfoundation/hardhat-toolbox": "^5.0.0",
    "@nomicfoundation/ignition-core": "^0.15.4",
    "@openzeppelin/contracts": "^5.0.2",
    "@typechain/ethers-v6": "^0.5.1",
    "ctf": "github:Cyfrin/ctf"
  }
}
```

### Contract Code

```solidity
// SPDX-License-Identifier: MIT
pragma solidity 0.8.24;

import {CTFRegistry} from "ctf/src/protocol/CTFRegistry.sol";

contract CyfrinSecurityCourseNft is CTFRegistry {
    constructor() CTFRegistry("Cyfrin Security Course NFT", "CSCN") {}
}
```

### Does this work on other EVMs? (If yes, please list at least 1 of them)

Yes. Arbitrum.

### Description of What Your Contract Does

It allows people to mint NFTs after solving solidity challenges

### Additional Details

So, typically, I don't recommend "just keep trying it again until it works", but that's what ended up solving this. I had gotten this to work on another similar script in the past by "just trying it again" and after like 6 tries of changing _absolutely nothing_ this script ended up working and the code deployed successfully.

Have others had issues where they get an error like this? 

```
Error: could not coalesce error (error={ "code": 3, "data": "0x", "message": "ErrorObject { code: ServerError(3), message: \"not enough gas to publish compressed bytecodes\", data: Some(RawValue(\"0x\")) }" }, payload={ "id": 27, "jsonrpc": "2.0", "method": "eth_sendRawTransaction", "params": [ 
```

**Comment by bxpana** [comment]
This might be due to not being able to fit in the block and by trying until it works if finally got all into a block. Will double check to see if that's the case

**Comment by mm-zk** [comment, marked as answer]
TL;DR - this was probably caused by sepolia's (L1) blob price spike. 
(@bxpana - we should add also a field (next to environment), where user can mention the current block height - would help with explaining issues like this).

Longer description:

ZKsync is a rollup, so our users have to pay a fee for pubdata that is then sent to L1. As L1 price (to be exact blob price) can be volatile, we model this volatility by changing the amount of gas that the L2 transaction has to pay when writing a byte to storage (or for example publishing the bytecode).

We rolled out the feature last week to `zks_getBlockDetails` that returns the current cost of pubdata publishing.

To show this on a concrete example:
```
curl --request POST \
  --url https://mainnet.era.zksync.io/ \
  --header 'Content-Type: application/json' \
  --data '{
      "jsonrpc": "2.0",
      "id": 1,
      "method": "zks_getBlockDetails",
      "params": [37534736]
    }'
    ```
    
   returns:
   ```
   ... 
    "fairPubdataPrice":4736052432,
    "l1BatchNumber":487660,
    "l1GasPrice":2960032769,
    "l1TxCount":0,
    "l2FairGasPrice":25000000,
    ```

This means that the the users pay 4'736'052'432 L1 wei per byte (so 4.7 Gwei). As the L2 gas price is 25'000'000 wei (0.025 Gwei) - this means that they pay 4.7 / 0.025 = 188 gas per byte published.

So what probably happened in your case, was a spike in this 'gas per byte' - which caused your transaction to run out of gas.

**Reply by PatrickAlphaC** [reply]
Got it, thanks!
